### PR TITLE
Add a post method that allows sending binary data

### PR DESCRIPTION
**Pull request by UVV-gh**
When the data to be sent comes from rather big file, it is more
efficient to read them from the file directly.

Providing data as a parameter to the exising post method could cause
extremely high memory usage.

Signed-off-by: Vyacheslav Yurkov <Vyacheslav.Yurkov@bruker.com>

### CODE REVIEW

**Comment by UVV-gh** [comment]
There are more details on the issue provided in this SO post https://stackoverflow.com/questions/54907281/robot-framework-send-binary-data-in-post-request-body-with

**Comment by lucagiove** [comment]
I don't know much about this part but isn't the `files` parameter works already in this way?
https://2.python-requests.org/en/master/user/quickstart/#post-a-multipart-encoded-file

**Comment by UVV-gh** [comment]
@lucagiove I assume that would be a different beast. The SO post explains the issue clearly. That would be a part of relevant request's documentation https://2.python-requests.org/en/master/user/advanced/#streaming-uploads

**Comment by lucagiove** [comment]
Sorry I missed to read the Stack Overflow I'll have a read.

**Comment by UVV-gh** [comment]
Looks like the patch is bit old now. I take a look at rebasing it...

**Comment by lucagiove** [comment]
Before working on it let me see if it's a feature that I would merge and in case how could be properly integrated I'll let you know soon.
There are some plan to change a lot the structure of the keywords, see the open feature issues.

**Comment by lucagiove** [comment]
Hi, I had a look at original requests library code when `files` parameter is passed and seems to open it as a file descriptor stream.
In any case another option to allow passing a file descriptor as `data` could be to create a keyword that returns a file descriptor. eg `Get File Descriptor` and use the standard requests keywords.
What do you think?

**Comment by UVV-gh** [comment]
Alright. I think that makes sense. Let me rebase onto latest master and update my pull request according to your suggestions.

**Comment by UVV-gh** [comment]
A few tests got broken I'm working on them..

Now I remember why I added a separate method for 'binary' post. The downsides are:
- The 'data' variable has to be checked in a few places if it's a file handler
- Without "with" the file handler effectively remains open until the end of the test (perhaps even until the end of all tests, I'm not sure when it gets released)

I hope these limitations are OK. I let you know when the tests are fixed

**Comment by UVV-gh** [comment]
I could, of course, add another keyword for closing a file handler.. not sure if it fits the design though :)

**Comment by UVV-gh** [comment]
Alright, I went the second way and close a handle right after POST request. Not sure what happens if exception is thrown on the way though.

The tests run. Please take a look and let me know what you think.

**Comment by lucagiove** [comment]
Hi! Thank you for your work!

About problem of multiple check of data: in the near future I want to remove all the 'data' parsing and pass directly to requests library, so this should be covered.
You can see my work in progress idea in this branch:
https://github.com/bulkan/robotframework-requests/tree/new-keywords
We can say that the file descriptor feature would work only for the "GET/POST./... On Session" keywords, I want to deprecate all the others.

Instead regarding the open file you're right since the library is global I think file descriptor would not be closed automatically (except if requests does it automatically).

**Comment by lucagiove** [comment]
I didn't check yet the code I was just answering to the previous comments.
Thanks I'll have a look to the PR soon hopefully.

File: src/RequestsLibrary/RequestsKeywords.py [hunk 1]
```diff
@@ -668,7 +669,7 @@ def post_request(
         ``timeout`` connection timeout
         """
         session = self._cache.switch(alias)
-        if not files:
+        if not files and not isinstance(data, io.IOBase):
```
Comment (lucagiove) [inline src/RequestsLibrary/RequestsKeywords.py]: I would add this check inside the `utils.format_data_according_to_header`

File: src/RequestsLibrary/RequestsKeywords.py [hunk 2]
```diff
@@ -683,6 +684,10 @@ def post_request(
             headers=headers,
             allow_redirects=redir,
             timeout=timeout)
+
+        if isinstance(data, io.IOBase):
+            data.close()
```
Comment (lucagiove) [inline src/RequestsLibrary/RequestsKeywords.py]: nice trick to make sure the file descriptor is closed

File: src/RequestsLibrary/log.py [hunk 3]
```diff
@@ -28,9 +29,12 @@ def log_request(
     args.pop('session', None)
     # This will log specific headers merged with session defined headers
     merged_headers = utils.merge_headers(session, args.pop('headers', None))
-    formatted_data = utils.format_data_to_log_string_according_to_headers(session,
-                                                                          args.pop('data', None),
-                                                                          merged_headers)
+    data = args.pop('data', None)
+    if data and not isinstance(data, io.IOBase):
```
Comment (lucagiove) [inline src/RequestsLibrary/log.py]: I would move this check inside `utils.format_data_to_log_string_according_to_headers`

File: tests/testcase.robot [hunk 4]
```diff
@@ -192,8 +193,18 @@ Post Request With Arbitrary Binary Data
     ${data}=  Get Binary File  ${CURDIR}${/}randombytes.bin
     &{headers}=  Create Dictionary  Content-Type=application/octet-stream   Accept=application/octet-stream
     ${resp}=  Post Request  httpbin  /post  data=${data}  headers=&{headers}
-    # TODO Compare binaries. Content is json with base64 encoded data
-    Log    "Success"
+    ${receivedData}=  Base64 Decode Data  ${resp.json()['data']}
```
Comment (lucagiove) [inline tests/testcase.robot]: Thanks for fixing the TODO!

File: src/RequestsLibrary/RequestsKeywords.py [hunk 5]
```diff
@@ -651,8 +652,8 @@ def post_request(
         ``data`` a dictionary of key-value pairs that will be urlencoded
                and sent as POST data
                or binary data that is sent as the raw body content
-               or passed as such for multipart form data if ``files`` is also
-                  defined
+               or passed as such for multipart form data if ``files`` is also defined
+               or file descriptor retreived by Get File Descriptor
```
Comment (lucagiove) [inline src/RequestsLibrary/RequestsKeywords.py]: Is this true only for POST?

File: src/RequestsLibrary/RequestsKeywords.py [hunk 6]
```diff
@@ -945,6 +950,10 @@ def _common_request(
 
         return resp
 
+    @staticmethod
+    def get_file_descriptor(path):
```
Comment (lucagiove) [inline src/RequestsLibrary/RequestsKeywords.py]: I think this needs some documentation

File: src/RequestsLibrary/RequestsKeywords.py (hunk 5 above)
Comment (UVV-gh) [reply src/RequestsLibrary/RequestsKeywords.py]: Alright, according to the docs it should also be PATCH and PUT

**Comment by UVV-gh** [comment]
Updated according to the comments

File: src/RequestsLibrary/RequestsKeywords.py (hunk 5 above)
Comment (lucagiove) [reply src/RequestsLibrary/RequestsKeywords.py:656]: in theory requests lib let you pass data to all method but I think it's fine for the moment

**Comment by lucagiove** [comment]
if you give me write access to the branch I'd like to add some small changes before merging

**Comment by UVV-gh** [comment]
Never did that before. Sent you an Email. Please check if that would be enough

**Comment by lucagiove** [comment]
@UVV-gh I was wandering how to explain the difference between using the file descriptor and `files` parameter in the documentation of the keywords.
You mentioned something about the multipart.

**Comment by UVV-gh** [comment]
I think multipart is related to "files" only. File descriptor is used for continuous data stream.

**Comment by lucagiove** [comment]
> Before merging I want to do a test comparing file descriptor and files usage.

Not needed, trying to rename and add more doc.

**Comment by lucagiove** [comment]
Renamed `Get File Descriptor` to `Get File For Streaming Upload` so that it's more clear what it has been made for instead of what it does.
@UVV-gh what do you think?
Updated also the doc to make clear the usage, maybe an example in the doc might help?

**Comment by UVV-gh** [comment]
Looks fine for me. As for the docs, it's up to you. Looking briefly through the documentation, I didn't find much of the description of other features. On the other hand, unit tests cover most of the use cases ;)

**Pull request merged**